[expo-go][ios] add logging of manifest exception

chrfalch opened the pull request:
# Why

When manifest exceptions happens in Expo Go we aren't starting/loading the bundle in a JS context, meaning we don't log the errors to the packager - they're caught and displayed in the ErrorView:

<img width="300" alt="image" src="https://github.com/user-attachments/assets/61cd7716-1bf2-496b-b099-6f01f4098523" />

Resolves ENG-15442 together with #36171

# How

This commit fixes this by introducing a helper class that can log to the packager without a JS context in Expo Go:

- Added `Exponent/Kernel/DevSupport/EXPackagerLogHelper.m/h` for connecting and logging to the packager. It logs a single message using a static method and then closes the web socket.
- Added call to log the error inside `Exponent/Kernel/Views/ErrorView.m` when the error-prop is updated

<img width="861" alt="image" src="https://github.com/user-attachments/assets/c97ba4fe-6a1b-4a0c-bf58-95d152621637" />

# Test Plan

- Run Expo Go and try to open a running app with SDK-52 against `main`
- Observe that Expo Go displays and error and that the same error is displayed in the CLI
 
# Checklist

- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

chrfalch committed:
[ios][expo-go] add logging of manifest exceptions

When manifest exceptions happens in Expo Go we aren't starting/loading the bundle in a JS context, meaning we don't log errors to the packager - just displaying the ErrorView.

This commit fixes this by introducing a helper class that can log to the packager without a JS context in Expo Go.

chrfalch committed:
minor adjustments

Kudo commented on apps/expo-go/ios/Exponent/Kernel/DevSupport/EXPackagerLogHelper.m:
```diff
@@ -0,0 +1,164 @@
+  EXLogLevelError
+};
+
+@interface EXPackagerLogHelper () <SRWebSocketDelegate>
```
is it possible to write the new `PackagerLogHelper` in swift rather than objective-c?

chrfalch replied:
Since it uses SocketRocket converting to swift fails :(

Kudo replied:
what specific error there? even though SocketRocket doesn't define as a module, we can still add it to the bridging header. we are adding code to expo-go app target, it's fine to use bridging header.

chrfalch replied:
Sorry - this works, just me being a bit cross-eyed. I've updated the PR.

brentvatne commented:
for the urls, let's do:

```
Learn how to upgrade to SDK 53: https://...
Learn how to install Expo Go for SDK 52: https://...
```

chrfalch commented:
> for the urls, let's do:
> 
> ```
> Learn how to upgrade to SDK 53: https://...
> Learn how to install Expo Go for SDK 52: https://...
> ```

These are built in a way that requires a bit more refactoring to the exception. Could we get this one through and do the refactoring of the `ManifestException` in a different PR?

chrfalch committed:
[expo-go][ios] converted obj-c to swift

chrfalch committed:
[expo-go][ios] swift lint

chrfalch committed:
swiftlint

chrfalch committed:
swiftlint

Kudo commented on apps/expo-go/ios/Exponent/Kernel/DevSupport/EXPackagerLogHelper.swift:
```diff
@@ -0,0 +1,145 @@
+
+    socket?.delegate = self
+
+    let queue = DispatchQueue(label: "host.exp.Exponent.EXPackagerLogHelper")
```
could we reuse some default DispatchQueue.global than having a new queue?

chrfalch replied:
I was under the impression that it is common to suggest that network should be run on a dedicated queue? in RN's RCTPackagerConnection we use the `RCTReconnectingWebSocket` and it is given its own queue.

chrfalch replied:
Good point. Have change the implementation a bit using an `OperationQueue`.

Kudo commented on apps/expo-go/ios/Exponent/Kernel/DevSupport/EXPackagerLogHelper.swift:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  deinit {
+    if let socket = socket, connected {
```
should we also cleanup socket even in disconnected state? given this is in the deinit state

chrfalch replied:
Something like this?
```suggestion
    socket?.close()
    self.socket = nil
    connected = false
```

Kudo commented on apps/expo-go/ios/Exponent/Kernel/DevSupport/EXPackagerLogHelper.swift:
```diff
@@ -0,0 +1,145 @@
+
+import Foundation
+import React
+
```
maybe rename this class a little bit. the name looks too generic for other purposes

chrfalch replied:
Do you have a suggestion? :)

chrfalch replied:
Ah, of course - I agree - I thought you meant the whole class :)

chrfalch committed:
codereview

Fixed deinit implementation to free socket better

chrfalch committed:
codereview: renamed EXLogLevel

alanjhughes commented on apps/expo-go/ios/Exponent/Kernel/DevSupport/EXPackagerLogHelper.swift:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    let type: String
+    switch logLevel {
+    case .info:
+      type = "info"
+    case .warning:
+      type = "warn"
+    case .error:
+      type = "error"
+    }
```
We could make the EXPackagerLogLevel cases match these strings and drop this switch statement. Then return `"level":  logLevel.rawValue`,

chrfalch replied:
Good point. Updated!

chrfalch committed:
[expo-go][ios] added converter to enum

Instead of having to convert inline in the code, the enum now has its own converter function.

chrfalch committed:
swiftlint

chrfalch committed:
swiftlint

EvanBacon commented:
The link markdown is inverted here, the text should be first and the link second, e.g. `[foo](https://)` not `[https://](foo)`

chrfalch committed:
[expo-go][ios] codereview: changed to use OperationQueue

Changed to use OperationQueue for the websocket as per review comment.

chrfalch commented:
> The link markdown is inverted here, the text should be first and the link second, e.g. `[foo](https://)` not `[https://](foo)`

Correct :) I'll try to fix this when I refactor the ManifestException class a bit later!
